File: Core/Models/PerformanceEntry.swift
import Foundation

public struct WellnessEntry: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var athleteID: EntityID
    public var recordedAt: Date
    public var sleepHours: Double
    /// 1...10 — higher = better mood.
    public var mood: Int
    /// 1...10 — higher = more sore.
    public var soreness: Int
    /// 1...10 — higher = more motivated.
    public var motivation: Int
    /// 1...10 — higher = more stressed.
    public var stress: Int
    /// 1...10 — Borg-style RPE for the most recent training session.
    public var rpePreviousSession: Int
    public var notes: String?

    public init(
        id: EntityID = UUID(),
        athleteID: EntityID,
        recordedAt: Date,
        sleepHours: Double,
        mood: Int,
        soreness: Int,
        motivation: Int = 5,
        stress: Int = 5,
        rpePreviousSession: Int,
        notes: String? = nil
    ) {
        self.id = id
        self.athleteID = athleteID
        self.recordedAt = recordedAt
        self.sleepHours = sleepHours
        self.mood = max(1, min(10, mood))
        self.soreness = max(1, min(10, soreness))
        self.motivation = max(1, min(10, motivation))
        self.stress = max(1, min(10, stress))
        self.rpePreviousSession = max(1, min(10, rpePreviousSession))
        self.notes = notes
    }

    // Decoding fallback so rows pre-Pillar-5 (no motivation/stress columns)
    // still load.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(EntityID.self, forKey: .id)
        self.athleteID = try c.decode(EntityID.self, forKey: .athleteID)
        self.recordedAt = try c.decode(Date.self, forKey: .recordedAt)
        self.sleepHours = try c.decode(Double.self, forKey: .sleepHours)
        self.mood = try c.decode(Int.self, forKey: .mood)
        self.soreness = try c.decode(Int.self, forKey: .soreness)
        self.motivation = try c.decodeIfPresent(Int.self, forKey: .motivation) ?? 5
        self.stress = try c.decodeIfPresent(Int.self, forKey: .stress) ?? 5
        self.rpePreviousSession = try c.decode(Int.self, forKey: .rpePreviousSession)
        self.notes = try c.decodeIfPresent(String.self, forKey: .notes)
    }
}

File: flutter_app/lib/core/models/performance_entry.dart
import 'entity_id.dart';

/// Port of Core/Models/PerformanceEntry.swift.

class WellnessEntry {
  final EntityID id;
  final EntityID athleteId;
  final DateTime recordedAt;
  final double sleepHours;

  /// 1…10 — higher = better mood.
  final int mood;

  /// 1…10 — higher = more sore.
  final int soreness;

  /// 1…10 — higher = more motivated.
  final int motivation;

  /// 1…10 — higher = more stressed.
  final int stress;

  /// 1…10 — Borg-style RPE for the most recent training session.
  final int rpePreviousSession;

  final String? notes;

  WellnessEntry({
    EntityID? id,
    required this.athleteId,
    required this.recordedAt,
    required this.sleepHours,
    required int mood,
    required int soreness,
    int motivation = 5,
    int stress = 5,
    required int rpePreviousSession,
    this.notes,
  }) : id = id ?? newEntityId(),
       mood = mood.clamp(1, 10),
       soreness = soreness.clamp(1, 10),
       motivation = motivation.clamp(1, 10),
       stress = stress.clamp(1, 10),
       rpePreviousSession = rpePreviousSession.clamp(1, 10);

  /// Backward-compat: rows pre-Pillar-5 (no motivation/stress columns) still
  /// load — missing fields fall back to 5.
  factory WellnessEntry.fromJson(Map<String, dynamic> json) => WellnessEntry(
    id: json['id'] as String,
    athleteId: json['athleteID'] as String,
    recordedAt: DateTime.parse(json['recordedAt'] as String),
    sleepHours: (json['sleepHours'] as num).toDouble(),
    mood: json['mood'] as int,
    soreness: json['soreness'] as int,
    motivation: json['motivation'] as int? ?? 5,
    stress: json['stress'] as int? ?? 5,
    rpePreviousSession: json['rpePreviousSession'] as int,
    notes: json['notes'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'athleteID': athleteId,
    'recordedAt': recordedAt.toIso8601String(),
    'sleepHours': sleepHours,
    'mood': mood,
    'soreness': soreness,
    'motivation': motivation,
    'stress': stress,
    'rpePreviousSession': rpePreviousSession,
    'notes': notes,
  };
}
